android/java/org/chromium/chrome/browser/brave_news/CardBuilderFeedCard.java

```diff
@@ -0,0 +1,1412 @@
+            DisplayAd displayAd = mNewsItem.getDisplayAd();
+            if (feedItemsCard == null) {
+                return null;
+            }
+            FeedItemCard feedItemCard = feedItemsCard.get(index);
+            FeedItem feedItem = feedItemCard.getFeedItem();
+
+            // FeedItem feedItem = mNewsItem.getFeedItems().get(index).getFeedItem();
+            switch (feedItem.which()) {
+                case FeedItem.Tag.Article:
+
+                    Article article = feedItem.getArticle();
+                    FeedItemMetadata articleData = article.data;
+                    itemData = article.data;
+
+                    break;
+                case FeedItem.Tag.PromotedArticle:
+                    PromotedArticle promotedArticle = feedItem.getPromotedArticle();
+                    FeedItemMetadata promotedArticleData = promotedArticle.data;
+                    mCreativeInstanceId = promotedArticle.creativeInstanceId;
+                    itemData = promotedArticle.data;
+
+                    mIsPromo = true;
+                    break;
+                case FeedItem.Tag.Deal:
+                    Deal deal = feedItem.getDeal();
+                    FeedItemMetadata dealData = deal.data;
+                    mOffersCategory = deal.offersCategory;
+                    itemData = deal.data;
+
+                    break;
+            }
+        } catch (Exception e) {
+            Log.e(TAG, " exceptionscaught getItemData e:" + e.getMessage());
+        }
+
+        final FeedItemMetadata itemDataFinal = itemData;
+
+        return itemDataFinal;
+    }
+
+    private void setListeners(
+            View view, String urlString, String creativeInstanceId, boolean isPromo) {
+        DisplayAd displayAd = mNewsItem.getDisplayAd();
+        view.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                if (isPromo) {
+                    // Updates the no. of promotion cards visited
+                    mBraveNewsController.onPromotedItemVisit(
+                            mNewsItem.getUuid(), creativeInstanceId);
+                } else if (displayAd != null) {
+                    // Updates the no. of ads cards visited
+                    mBraveNewsController.onDisplayAdVisit(
+                            displayAd.uuid, displayAd.creativeInstanceId);
+                } else {
+                    // Brave News updates the no. of "normal" cards visited
+                    int visitedNewsCardsCount = SharedPreferencesManager.getInstance().readInt(
+                            BravePreferenceKeys.BRAVE_NEWS_CARDS_VISITED);
+                    visitedNewsCardsCount++;
+                    SharedPreferencesManager.getInstance().writeInt(
+                            BravePreferenceKeys.BRAVE_NEWS_CARDS_VISITED, visitedNewsCardsCount);
+                    if (visitedNewsCardsCount % 4 == 0) {
```

Comment:
For visits, I think we should call the service whenever this changes, since it's infrequent and we want to know when it's 1, which is going to be the most common apart from 0.

Reply:
ok, so i'll use `if (visitedNewsCardsCount > 0)` then